File: First-Edition/BookSamples/rosetta-stone/rust/rsc2p2p1_type_level_literals/src/assets/gol.rs
use std::marker::PhantomData;
use crate::automaton;

use automaton::Board;
use automaton::CellWorld;
use automaton::Cell;
use automaton::IAutomaton;

pub enum GoLRule {}

pub type GoL = CellWorld<GoLRule>;

fn gol_step(board: &Board) -> Board {
  let mut new_board = board.clone();

  for (i, row) in board.iter().enumerate() {
    for (j, cell) in row.iter().enumerate() {
      let nh = automaton::get_neighbor_alive(i, j, &board);

      new_board[i][j] = match (cell, nh) {
        (Cell::Alive, 2) => Cell::Alive,
        (Cell::Alive, 3) => Cell::Alive,
        (Cell::Dead,  3) => Cell::Alive,
        _ => Cell::Dead,
      }
    }
  }

  return new_board;
}

pub fn make_gol(board: &Board) -> GoL {
  return CellWorld::CW{ board: board.to_vec(), _marker: PhantomData::<GoLRule> };
}

impl IAutomaton for GoL {
  fn step(&self) -> Self {
    let new_board = match self {
      CellWorld::CW {board, ..} => gol_step(&board)
    };
    return CellWorld::CW{ board: new_board, _marker: PhantomData::<GoLRule> };
  }
}

File: First-Edition/BookSamples/rosetta-stone/scala3/rsc2p1p3_traits_oop/src/main/scala/assets/GoL.scala
import Automaton.*

object GoL:

  case class GoL(board: Board) extends IAutomaton {
    def step(): GoL =
      val newBoard = board.indices.map { x =>
        board(x).indices.map { y =>
          nextState(board, x, y)
        }.toVector
      }.toVector
      GoL(newBoard)
    def unwrap(): Board = board
  }

  def nextState(board: Board, x: Int, y: Int): Cell = {
    val neighbors = for {
      i <- -1 to 1
      j <- -1 to 1
      if i != 0 || j != 0
      if board.isDefinedAt(x + i) && board(x + i).isDefinedAt(y + j)
    } yield board(x + i)(y + j)

    val aliveNeighbors = neighbors.count(_ == Alive)

    board(x)(y) match {
      case Alive if aliveNeighbors < 2 || aliveNeighbors > 3 => Dead
      case Alive => Alive
      case Dead if aliveNeighbors == 3 => Alive
      case Dead => Dead
    }
  }

